data/maps/SP_Scenarios/Along_The_River.wmf/scripting/functions.lua

```diff
@@ -0,0 +1,119 @@
+      end
+   end
+end
+
+-- =======================================================================
+
+function immovable_has_name(immovable, name)
+   return (immovable ~= nil and immovable.descr.name == name)
+end
+
+-- =======================================================================
+
+function add_wares(point, ware, amount)
+   local immovable = point.immovable
+   if immovable then
+      if immovable.descr.type_name == "flag" or immovable.descr.type_name == "warehouse" or immovable.descr.type_name == "productionsite" then
+         immovable:set_wares(ware, immovable:get_wares(ware) + amount)
+      end
+   end
+end
+
+-- =======================================================================
+
+-- Transports wares to flag, warehouse or productionsite.
+-- Waits if the building does not exist.
+function transport_wares(player, point, ware, amount, alert, image)
+   local count = amount
+   while count > 0 do
+      sleep(499)
+      local immovable = point.immovable
+      if immovable then
+         if immovable.descr.type_name == "flag" or immovable.descr.type_name == "warehouse" or immovable.descr.type_name == "productionsite" then
+            immovable:set_wares(ware, immovable:get_wares(ware) + 1)
+            count = count - 1
+         end
+      end
+   end
+   if(alert) then
+      local ware_name =  wl.Game():get_ware_description(ware).descname
+      local immovable = point.immovable
+      if immovable then
+         if immovable.descr.type_name == "flag" or immovable.descr.type_name == "warehouse" or immovable.descr.type_name == "productionsite" then
+            if image then
+               send_to_inbox(player, _"Transport", li_image(image, ngettext("%1% unit of %2% has been brought to your ‘%3%’ building.", "%1% units of %2% have been brought to your ‘%3%’ building.", amount):bformat(amount, ware_name, immovable.descr.descname)), { field = point, popup = true, })
+            else
+               send_to_inbox(player, _"Transport", p(ngettext("%1% unit of %2% has been brought to your ‘%3%’ building.", "%1% units of %2% have been brought to your ‘%3%’ building.", amount):bformat(amount, ware_name, immovable.descr.descname)), { field = point, popup = true, })
+            end
+         end
+      end
+   end
+end
+
+-- =======================================================================
+
+-- Spawns soldiers in a warehouse, militarysite or trainingsite with intervals in ms.
+-- e.g.: spawn_soldiers(map:get_field(hq_x,hq_y), {3,5,0,2}, 45, 4999)
+-- Stops if the object does not exist or is invalid.
+function spawn_soldiers(point, ranking, count, interval)
+   while count > 0 do
+      sleep(interval)
+      local immovable = point.immovable
+      if immovable then
+         if immovable.descr.type_name == "warehouse" or immovable.descr.type_name == "militarysite" or immovable.descr.type_name == "trainingsite" then
+            immovable:set_soldiers(ranking, immovable:get_soldiers(ranking) + 1)
```

Comment:
This will delete every soldier in the building who does not match `ranking`. Intentional?

Reply:
in this case I guess no. so first get all soldiers and then iterate over the descriptions and just add to the matching descritption right?

Reply:
i was trying this code but for reasons i dont't know it doesn't work

            local soldiers = immovable:get_soldiers("all")
            soldiers[ranking] = immovable:get_soldiers(ranking) + 1
            immovable:set_soldiers(soldiers)

Reply:
Tables are treated as references to (unique) objects in Lua, so if you create a local table `ranking` it is treated as an object different from the table that is used as a key in `soldiers`. Something like this should work:
```
local did_set = false
for key,value in pairs(soldiers) do
   if key[1] == ranking[1] and key[2] == ranking[2] and key[3] == ranking[3] and key[4] == ranking[4] then
      soldiers[key] = value + 1
      did_set = true
      break
   end
end
if not did_set then soldiers[ranking] = 1 end
immovable:set_soldiers(soldiers)
```

---

In case the explanation above is not clear… When Lua compares two tables, it actually compares their memory addresses. So your table actually looks like
```
soldiers = {
   0x123456 = 5,
   0x234567 = 3,
}
```
The expression `soldiers[ranking]` evaluates to `soldiers[0x987abc]`, and this key does not exist (is `nil`). A `for … in pairs` allows you to access the original pairs `(0x123456, 5), (0x234567, 3)`.
Hope this is a bit clearer…

Reply:
Well I tested it with the HQ of player one sometimes it added a soldier sometimes it did not.  that was what I didn't get.  
I also checked via debug messages the value before the setting line and the value after. sometimes they differd by 1 sometimes not.  So I don't get what was wrong with my code. What I wanted to achieve is that if the key doesn't exist it should be added. 
to make this more plausible I don't get why 

"soldiers[ranking] = int"

in my code does not work every time while  

"if not did_set then soldiers[ranking] = 1 end "

in your code should work

Reply:
The code I posted above (untested!) first tries to locate an equivalent key in the table and replace its value with the updated one. If this fails (`did_set == false`) then a *new* key is added to the table, which will work because at this time we know that `soldiers` does not contain a key semantically equivalent to `ranking`.

Let's look what happens in your code. The initial state is
```
soldiers = {
   0x123456 <0,0,0,0> = 5,
   0x234567 <3,5,0,2> = 3,
}
ranking = 0x987abc <3,5,0,2>
```
Now we do `soldiers[ranking] = <expression that evaluates to 4>` and we get the updated table
```
soldiers = {
   0x123456 <0,0,0,0> = 5,
   0x234567 <3,5,0,2> = 3,
   0x987abc <3,5,0,2> = 4,
}
```
When the C++ interface parses a table, it is always undefined in which order the table's keys are iterated and interpreted. So in this case, either the table entry `0x234567 = 3` silently overwrites the interpretation of the table entry `0x987abc = 4`, or the other way around – so it's a random 50/50 chance which of the two values is actually used by the engine in the end.

Whereas, in my code the assignment inside the `for` loop looks like:
```
key = 0x234567 <3,5,0,2>
value = 3
soldiers[0x234567 <3,5,0,2>] = 3 + 1
```
The key to which we assign here is already present in the table, so it's replaced, and `soldiers` then looks like this:
```
soldiers = {
   0x123456 <0,0,0,0> = 5,
   0x234567 <3,5,0,2> = 4,
}
```
which should unambigously give the correct result.

Reply:
ok that explains it. Thanks for explaining and the patience